mayastor/src/bdev/nvmx/controller.rs

```diff
@@ -843,19 +850,17 @@ pub extern "C" fn nvme_poll_adminq(ctx: *mut c_void) -> i32 {
     let result = context.process_adminq();
 
     if result < 0 {
-        error!(
-            "process adminq: {}: {}",
-            context.name,
-            Errno::from_i32(result.abs())
-        );
         if context.start_device_destroy() {
-            info!("dispatching device destroy: {}", context.name);
+            error!(
+                "process adminq: {}: {}",
+                context.name,
+                Errno::from_i32(result.abs())
+            );
+            info!("dispatching nexus fault and retire: {}", context.name);
             let dev_name = context.name.to_string();
             Reactors::master().send_future(async move {
-                let _ = destroy_device(dev_name).await;
+                nexus::nexus_fault_child(dev_name).await;
```

Comment:
I am a bit concerned that controller has knowledge about the holding nexus directly.
Can't we rather add indirection here by utilising existing event mechanism in NVMx controller to notify all the listeners about device failure ?
Nexus already registers itself as a listener to monitor events on its children devices, so we could probably extend the event handler to handle also device I/O failures.

Reply:
Actually it doesn't, that function is iterating through all nexuses, and their children, looking for a child whose device name matches this controller name.

Reply:
The idea behind the controller event mechanism was to decouple controller events from the logic that processes it.
This situation with faulting nexus child looks like a very good example of that. We might want to have more than one consumer of such an event potentially.

Reply:
I had a look at the `DeviceEventType::DeviceRemoved` issued at the end of `destroy_device()` (i.e. the previous fix for this issue) and this event is sent after the device is removed in order for the child to respond to it, in this case by removing the child from the nexus.
I could introduce a new event type of `FaultChild` then move `nexus::nexus_fault_child()` into `child_event_listener()` so the faulting is handled as an event instead of sending a `Future`.

Reply:
Sounds great ! Thanks !